File: doctor/src/main/java/com/sumian/sddoctor/h5/BaseWebViewFragment.java
package com.sumian.sddoctor.h5;

import android.view.View;

import com.blankj.utilcode.util.LogUtils;
import com.sumian.common.h5.widget.SWebView;
import com.sumian.common.h5.widget.SWebViewLayout;
import com.sumian.sddoctor.BuildConfig;
import com.sumian.sddoctor.R;
import com.sumian.sddoctor.app.AppManager;
import com.sumian.sddoctor.base.BaseFragment;
import com.sumian.sddoctor.widget.TitleBar;
import com.tencent.smtt.sdk.WebView;


/**
 * Created by sm
 * on 2018/5/25 10:03
 * desc:
 **/
@SuppressWarnings("ConstantConditions")
public abstract class BaseWebViewFragment extends BaseFragment implements TitleBar.OnBackClickListener, SWebView.OnWebViewListener {

    protected TitleBar mTitleBar;

    protected View mDivider;

    protected SWebViewLayout mSWebViewLayout;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_main_base_webview;
    }

    @Override
    public void initWidget(View root) {
        super.initWidget(root);
        mTitleBar = root.findViewById(R.id.title_bar);
        mTitleBar.setOnBackClickListener(this);
        mDivider = root.findViewById(R.id.view_divider);
        mSWebViewLayout = root.findViewById(R.id.sm_webview_container);
        mSWebViewLayout.loadRequestUrl(getCompleteUrl());
//        new H5LocalCacheInterceptor(getActivity(), H5LocalCachePath.ASSET_PATH).interceptH5Request(mSWebViewLayout.getSWebView());
    }

    @Override
    public void initData() {
        super.initData();
        registerHandler(mSWebViewLayout.getSWebView());
        mSWebViewLayout.setWebListener(this);
    }

    @Override
    public void onResume() {
        super.onResume();
        mSWebViewLayout.resumeWebView();
    }

    @Override
    public void onPause() {
        super.onPause();
        mSWebViewLayout.pauseWebView();
    }

    @Override
    public void onDestroy() {
        mSWebViewLayout.destroyWebView();
        super.onDestroy();
    }

    protected String h5HandlerName() {
        return null;
    }

    protected void registerHandler(SWebView sWebView) {

    }

    @Override
    public void onBack(View v) {
        if (!mSWebViewLayout.webViewCanGoBack()) {
            getActivity().onBackPressed();
        }
    }

    protected String getCompleteUrl() {
        StringBuilder stringBuilder = new StringBuilder();
        String urlServerPart = getUrlServerPart();
        String urlContentPart = getUrlContentPart();
        boolean contentContainerParams = urlContentPart.contains("?");
        stringBuilder.append(urlServerPart)
                .append(urlContentPart)
                .append(contentContainerParams ? "&" : "?")
                .append(getUrlToken());
        String url = stringBuilder.toString();
        LogUtils.d("load url: %s", url);
        return url;
    }

    protected String getUrlServerPart() {
        return BuildConfig.BASE_H5_URL;
    }

    protected String getUrlContentPart() {
        return null;
    }

    private String getUrlToken() {
        return "token=" + AppManager.getAccountViewModel().getToken();
    }

    @Override
    public void onReceiveTitle(WebView webView, String title) {
        mTitleBar.setTitle(title);
    }

    @Override
    public void onPageStarted(WebView webView) {

    }

    @Override
    public void onProgressChange(WebView webView, int i) {

    }

    @Override
    public void onPageFinish(WebView webView) {

    }

    @Override
    public void onRequestErrorCallback(WebView webView, int i) {

    }

    @Override
    public void onRequestNetworkErrorCallback(WebView webView) {

    }
}

File: common/src/main/java/com/sumian/common/h5/BaseWebViewFragment.kt
package com.sumian.common.h5

import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.core.view.isVisible
import com.blankj.utilcode.util.LogUtils
import com.google.gson.reflect.TypeToken
import com.sumian.common.R
import com.sumian.common.base.BaseViewModel
import com.sumian.common.base.BaseViewModelFragment
import com.sumian.common.dialog.SumianImageTextDialog
import com.sumian.common.h5.bean.H5ShowToastData
import com.sumian.common.h5.bean.ShareData
import com.sumian.common.h5.widget.SWebView
import com.sumian.common.h5.widget.SWebViewLayout
import com.sumian.common.utils.JsonUtil
import com.sumian.common.utils.StatusBarUtil
import com.tencent.smtt.sdk.WebView
import kotlinx.android.synthetic.main.base_h5_fragment.*
import kotlinx.android.synthetic.main.base_h5_fragment.sm_webview_container

abstract class BaseWebViewFragment : BaseViewModelFragment<BaseViewModel>(), SWebView.OnWebViewListener {

    protected var mSumianImageTextDialog: SumianImageTextDialog? = null
    protected var mShare: ImageView? = null
    protected var mTitleBar: LinearLayout? = null

    override fun onPageStarted(view: WebView?) {
    }

    override fun onProgressChange(view: WebView?, newProgress: Int) {
        if (newProgress <= 20) {
            iv_share?.visibility = View.INVISIBLE
        }
    }

    override fun onPageFinish(view: WebView?) {
    }

    override fun onRequestErrorCallback(view: WebView?, responseCode: Int) {
    }

    override fun onRequestNetworkErrorCallback(view: WebView?) {
    }

    override fun onReceiveTitle(view: WebView?, title: String?) {
        tv_title?.text = title
    }

    override fun getLayoutId(): Int {
        return R.layout.base_h5_fragment
    }

    open fun getSWebViewLayout(): SWebViewLayout? {
        return sm_webview_container
    }

    override fun initWidget() {
        super.initWidget()
        mShare = iv_share
        mTitleBar = fl_header
    }

    override fun initData() {
        super.initData()
        getSWebViewLayout()?.setWebListener(this)
        getSWebViewLayout()?.loadRequestUrl(getCompleteUrl())
        registerHandler(getSWebViewLayout()!!.sWebView)
        registerBaseHandler(getSWebViewLayout()!!.sWebView)
        iv_back?.setOnClickListener {
            if (sm_webview_container.sWebView.canGoBack()) {
                sm_webview_container.sWebView.goBack()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        sm_webview_container?.resumeWebView()
    }

    override fun onPause() {
        sm_webview_container?.pauseWebView()
        super.onPause()
    }

    protected open fun registerHandler(sWebView: SWebView) {}

    private fun registerBaseHandler(sWebView: SWebView) {
        sWebView.registerHandler("showToast") { data, function ->
            LogUtils.d(data)
            val toastData = H5ShowToastData.fromJson(data)
            if (mSumianImageTextDialog != null) {
                mSumianImageTextDialog!!.dismiss()
            } else {
                mSumianImageTextDialog = SumianImageTextDialog(activity!!)
            }
            mSumianImageTextDialog!!.show(toastData)
        }
        sWebView.registerHandler("hideToast") { data, function ->
            LogUtils.d(data)
            val toastData = H5ShowToastData.fromJson(data)
            if (mSumianImageTextDialog != null) {
                mSumianImageTextDialog!!.dismiss(toastData.delay)
            }
        }
        sWebView.registerHandler("return") { data, function -> goBack() }
        sWebView.registerHandler("updatePageUI") { data, function ->
            val map = JsonUtil.fromJson<Map<String, Any>>(data, object : TypeToken<Map<String, Any>>() {
            }.type) ?: return@registerHandler
            for ((key, value) in map) {
                when (key) {
                    "showNavigationBar" -> if (value is Boolean) {
                        fl_header?.isVisible = value
                    }
                    "showTitle" -> if (value is Boolean) {
                        tv_title?.isVisible = value
                    }
                    "showBackArrow" -> if (value is Boolean) {
                        iv_back?.visibility = if (value) View.VISIBLE else View.INVISIBLE
                    }
                    "setStatusBarTextColorDark" -> if (value is Boolean) {
                        if (activity == null) {
                            return@registerHandler
                        }
                        StatusBarUtil.setStatusBarTextColorDark(activity!!, value)
                    }
                    else -> {
                    }
                }
            }
        }
        sWebView.registerHandler("goToPage") { data, function ->
            run {
                val page = JsonUtil.getJsonObject(data)?.get("page")?.asString
                        ?: return@registerHandler
                onGoToPage(page, data)
            }
        }
    }

    protected open fun onGoToPage(page: String, rawData: String) {

    }

    protected open fun onShare(shareData: ShareData) {

    }

    protected open fun getCompleteUrl(): String {
        val stringBuilder = StringBuilder()
        val urlServerPart = getUrlServerPart()
        val urlContentPart = getUrlContentPart()
        val contentContainerParams = urlContentPart != null && urlContentPart.contains("?")
        stringBuilder.append(urlServerPart)
                .append(if (urlContentPart.isNullOrEmpty()) "" else urlContentPart)
                .append(if (contentContainerParams) "&" else "?")
                .append(getUrlToken())
        val url = stringBuilder.toString()
        LogUtils.d("load url: %s", url)
        return url
    }

    protected fun getUrlServerPart(): String {
        return WebViewManger.getInstance().getBaseUrl() ?: ""
    }

    protected open fun getUrlContentPart(): String? {
        return null
    }

    private fun getUrlToken(): String {
        return "token=" + getToken()
    }

    protected fun getToken(): String? {
        return WebViewManger.getInstance().getToken()
    }

    protected fun reload() {
        getSWebViewLayout()?.reload()
    }

    protected fun getWebView(): SWebView? {
        return getSWebViewLayout()?.sWebView
    }

    fun goBack(): Boolean {
        return getSWebViewLayout()?.webViewCanGoBack() ?: false
    }

    override fun onDestroyView() {
        super.onDestroyView()
        sm_webview_container?.destroyWebView()
        mSumianImageTextDialog?.release()

    }

}